# Medidas de variabilidad


## 1. ¿Qué son las medidas de variabilidad (o de dispersión)?

> Son una manera de cuantificar **qué tanto se alejan los datos de su valor central**

## 2. ¿Cuáles son las medidas de variabilidad más usadas?

- Datos numéricos:
    - La desviación estándar
    - El rango intercuartiles
- Datos categóricos:
    - Usualmente no se mide la variabilidad. En el análisis univariado veremos formas de analizar la distribución de las diferentes categorías.

### 2.1. La desviación estándar

> Mide en promedio qué tanto se alejan los valores de la media de la variable.

Notación:
- $s$ cuando nos referimos a la desviación estándar de la muestra
- $\sigma$ cuando nos referimos a la desviacón estándar de la población

Matemáticamente se define como:

$$s = \sqrt{\frac{\sum_{i=1}^n (x_i-\bar{x})^2}{n-1}}$$

donde:

- $n$ es el número de datos que contiene la variable
- $\bar{x}$ es el promedio de la variable
- $x_i$ es cada uno de los datos de la variable
- $n-1$ es el **número de grados de libertad** (esto lo entenderemos más adelante en próximos cursos)

Al igual que la media, la desviación estándar es una medida propia de la **distribución normal** así que se debería usar si nuestros datos numéricos tienen este tipo de distribución.

Ejemplo:

In [1]:
import numpy as np

arreglo1 = np.array([3, 5, 9, 8])
print(f'Media = {np.mean(arreglo1)}, s = {np.std(arreglo1)}')

Media = 6.25, s = 2.384848003542364


Lo anterior quiere decir que, en promedio, los datos anteriores se alejan 2.38 unidades de la media (que tiene un valor de 6.25).

Veamos este segundo ejemplo:

In [4]:
arreglo2 = np.array([-17, 5, 19, 18])
print(f'Media = {np.mean(arreglo2)}, s = {np.std(arreglo2)}')

Media = 6.25, s = 14.515078366994786


En este segundo caso vemos que la variable tiene el mismo valor promedio del primer ejemplo (6.25). Sin embargo, tiene una desviación estándar mucho más grande (14.5 vs. 2.38). Esto quiere decir que **en el segundo arreglo los datos están más dispersos que en el primero**.

Al igual que con la media, la desviación estándar también **es sensible a valores extremos**:

In [5]:

arreglo1 = np.array([3, 5, 9, 8])
arreglo2 = np.array([3, 5, 9, 8, 500])
print('Variable sin valores extremos:')
print(f'Media = {np.mean(arreglo1)}, s = {np.std(arreglo1)}')
print('-'*30)
print('Variable CON valores extremos:')
print(f'Media = {np.mean(arreglo2)}, s = {np.std(arreglo2)}')

Variable sin valores extremos:
Media = 6.25, s = 2.384848003542364
------------------------------
Variable CON valores extremos:
Media = 105.0, s = 197.51151865144473


Este efecto se puede minimizar usando el **rango intercuartiles**, una medida de variabilidad más robusta frente a la presencia de valores extremos.

### 2.2. El rango intercuartiles

El rango intercuartiles resulta muy útil para medir la dispersión de los datos cuando:

- Tenemos distribuciones normales o NO normales
- Tenemos valores extremos

Antes de definir el rango intercuartiles debemos entender el concepto de **percentil**:

> El percentil *P* es un valor dentro de la variable tal que al menos el *P%* de los valores están **por debajo** de dicho valor y el *(100-P)%* restante está **por encima** de este valor.

Entendamos esto con un sencillo ejemplo: creemos un arreglo con 10 valores numéricos. Imprimamos este arreglo en pantalla así como el mismo arreglo pero organizado de forma ascendente:

In [6]:
arreglo = np.array([5, 3, 12, 17, 25, 32, 15, 8, 10, 9])

print(f'Arreglo original: {arreglo}')
print(f'Arreglo organizado: {np.sort(arreglo)}')

Arreglo original: [ 5  3 12 17 25 32 15  8 10  9]
Arreglo organizado: [ 3  5  8  9 10 12 15 17 25 32]


¿Cuál sería el percentil 10?

- El percentil 10 sería el valor numérico (dentro del arreglo organizado) por debajo del cual se encuentra el 10% de los datos y por encima del cual está el otro 90%.
- Si tenemos 10 datos, el 10% de estos será 1 dato. Es decir que el percentil 10 será el valor numérico por debajo del cuál se encuentra 1 dato y por encima del cual se encuentran los 9 datos restantes.

Calculemos este percentil 10 con ayuda de NumPy:

In [8]:
# Cálculo del percentil 10
ps = [10, 40]
percentiles = np.percentile(arreglo, ps)

# Mostrar los percentiles calculados
print(f'Arreglo organizado: {np.sort(arreglo)}')
for p, pv in zip(ps, percentiles):
    print(f"Percentil {p}: {pv}")

Arreglo organizado: [ 3  5  8  9 10 12 15 17 25 32]
Percentil 10: 4.8
Percentil 40: 9.6


Para el arreglo anterior el percentil 10 es igual a 4.8. Esto quiere decir que:

- Por debajo de 4.8 tendremos el 10% de los datos.
- Por encima de 4.8 tendremos el 90% de los datos.

Si verificamos, en efecto por debajo de 4.8 (percentil 10), tendremos el valor 3 (1 dato). Mientras que por encima tendremos los datos 5, 8, 9, 10, 12, 15, 17, 25 y 32 (9 datos).

Existen tres percentiles de interés cuando analizamos la dispersión de los datos:

- Percentil 25, también llamado **primer cuartil**
- Percentil 50, también llamado **segundo cuartil** (que es exactamente igual a la mediana de los datos)
- Percentil 75, también llamado **tercer cuartil**

¡Y se denominan cuartiles pues dividen la variable en cuatro partes iguales!

Veamos un ejemplo de cálculo de estos cuartiles:

In [9]:
# Cálculo de cuartiles 1, 2 y 3 (percentiles 25, 50 y 75)

ps = [25, 50, 75]
cuartiles = np.percentile(arreglo, ps)

print(f'Arreglo organizado: {np.sort(arreglo)}')
for p, pv in zip(ps, cuartiles):
    print(f"Percentil {p}: {pv}")

Arreglo organizado: [ 3  5  8  9 10 12 15 17 25 32]
Percentil 25: 8.25
Percentil 50: 11.0
Percentil 75: 16.5


Y teniendo una idea clara de lo que son los percentiles y los cuartiles ya estamos listos para definir el rango intercuartiles:

> El *rango intercuartiles* (IQR por sus siglas en Inglés) de una variable es una medida de dispersión que se calcula como la diferencia entre el cuartil 3 y el cuartil 1

¿Y por qué es una medida de dispersión? Por que recordemos que entre el cuartil 3 y el 1 está el cuartil 2, que es la misma mediana. Así:

- Si el rango intercuartiles es "grande" quiere decir que los cuartiles 3 y 1 estarán "alejados" de la mediana: "alta" dispersión
- Si el rango intercuartiles es "pequeño" quiere decir que los cuartiles 3 y 1 estarán "más cerca" de la mediana: "baja" dispersión

Entendamos esto con un ejemplo sencillo:

In [10]:
# Creemos un primer arreglo, calculemos su rango intercuartiles y su mediana
arreglo = np.linspace(-5,5,10) # Arreglo con 10 valores entre -5 y 5

# Cálculo de cuartiles 1, 2 (mediana) y 3:
q1 = np.percentile(arreglo, 25)
q2 = np.percentile(arreglo, 50)
q3 = np.percentile(arreglo, 75)

# Rango intercuartiles
iqr = q3-q1

# Imprimir información en pantalla
print(f'Arreglo: {arreglo}')
print(f'Mediana: {q2}, IQR: {iqr}')

Arreglo: [-5.         -3.88888889 -2.77777778 -1.66666667 -0.55555556  0.55555556
  1.66666667  2.77777778  3.88888889  5.        ]
Mediana: 0.0, IQR: 5.000000000000001


Ahora creemos un segundo arreglo. En este caso tendrá la misma mediana pero los datos estarán más dispersos:

In [11]:
# Creemos un primer arreglo, calculemos su rango intercuartiles y su mediana
arreglo = np.linspace(-25,25,10) # Arreglo con 10 valores entre -25 y 25

# Cálculo de cuartiles 1, 2 (mediana) y 3:
q1 = np.percentile(arreglo, 25)
q2 = np.percentile(arreglo, 50)
q3 = np.percentile(arreglo, 75)

# Rango intercuartiles
iqr = q3-q1

# Imprimir información en pantalla
print(f'Arreglo: {arreglo}')
print(f'Mediana: {q2}, IQR: {iqr}')

Arreglo: [-25.         -19.44444444 -13.88888889  -8.33333333  -2.77777778
   2.77777778   8.33333333  13.88888889  19.44444444  25.        ]
Mediana: 0.0, IQR: 24.999999999999996


Vemos que ambos arreglos tienen un valor central (mediana) igual a cero. Sin embargo el primero tiene un IQR de 5 mientras que el segundo uno de casi 25. Es decir que los datos del segundo arreglo tienen más dispersión (se alejan más de la mediana) que el primero.

Y al igual que ocurría con la mediana, el rango intercuartiles también es menos sensible a la presencia de *outliers* en comparación con la desviación estándar:

In [12]:
# Primer arreglo: sin "outliers"
arreglo = np.linspace(-5,5,10) # Arreglo con 10 valores entre -5 y 5

# Cálculo de cuartiles 1, 2 (mediana) y 3:
q1 = np.percentile(arreglo, 25)
q2 = np.percentile(arreglo, 50)
q3 = np.percentile(arreglo, 75)

# Rango intercuartiles
iqr = q3-q1

# Imprimir información en pantalla
print(f'Arreglo SIN valores extremos: {arreglo}')
print(f'Mediana: {q2}, IQR: {iqr}')

Arreglo SIN valores extremos: [-5.         -3.88888889 -2.77777778 -1.66666667 -0.55555556  0.55555556
  1.66666667  2.77777778  3.88888889  5.        ]
Mediana: 0.0, IQR: 5.000000000000001


Ahora agreguemos al arreglo anterior un valor extremo de 500:

In [13]:
arreglo = np.linspace(-5,5,10) # Arreglo con 10 valores entre -5 y 5

# Agregar valor extremo
arreglo = np.append(arreglo, [500])

# Cálculo de cuartiles 1, 2 (mediana) y 3:
q1 = np.percentile(arreglo, 25)
q2 = np.percentile(arreglo, 50)
q3 = np.percentile(arreglo, 75)

# Rango intercuartiles
iqr = q3-q1

# Imprimir información en pantalla
print(f'Arreglo CON valores extremos: {arreglo}')
print(f'Mediana: {q2}, IQR: {iqr}')

Arreglo CON valores extremos: [ -5.          -3.88888889  -2.77777778  -1.66666667  -0.55555556
   0.55555556   1.66666667   2.77777778   3.88888889   5.
 500.        ]
Mediana: 0.5555555555555554, IQR: 5.555555555555556


Vemos que sí hay un cambio en la mediana (0.55 vs. 0) y en el IQRT (5.55 vs. 5.0) sin y con valores extremos, pero este cambio es mínimo a pesar del valor tan alto del *outlier*.

## 3. Ejemplo práctico

Pongamos en práctica todos estos conceptos. Leamos nuevamente el dataset `tasas_crimenes.csv` que contiene la tasa de asesinatos (por cada 100.000 habitantes) en diferentes estados de Estados Unidos y midamos la dispersión que tienen la población y la tasa de asesinatos, usando tanto la desviación estándar como el IQR.

Comencemos leyendo el set de datos:

In [14]:
# Leer dataset
import pandas as pd

ruta = '/Users/miguel/Library/CloudStorage/GoogleDrive-miguel@codificandobits.com/My Drive/02-CODIFICANDOBITS.COM/04-Academia/01-Cursos/21-2024-03-Estadística-Descriptiva/datos/'

df = pd.read_csv(ruta + 'tasas-crimenes.csv')
df

,State,Population,Murder.Rate,Abbreviation
0,Alabama,4779736,5.7,AL
1,Alaska,710231,5.6,AK
2,Arizona,6392017,4.7,AZ
3,Arkansas,2915918,5.6,AR
4,California,37253956,4.4,CA
5,Colorado,5029196,2.8,CO
6,Connecticut,3574097,2.4,CT
7,Delaware,897934,5.8,DE
8,Florida,18801310,5.8,FL
9,Georgia,9687653,5.7,GA


Para calcular la desviación estándar en Pandas simplemente podemos usar el método `std()`:

In [15]:
print('Población:')
print(f"\tMedia: {df['Population'].mean()}")
print(f"\tDesviación: {df['Population'].std()}")

print('='*30)
print('Tasa de asesinatos:')
print(f"\tMedia: {df['Murder.Rate'].mean()}")
print(f"\tDesviación: {df['Murder.Rate'].std()}")

Población:
	Media: 6162876.3
	Desviación: 6848235.347401142
Tasa de asesinatos:
	Media: 4.066
	Desviación: 1.9157361243029227


Para calcular el rango intercuartiles debemos implementar algunas líneas de código adicionales:

1. Calcular los cuartiles 1 y 3 con la función "quantile"
2. Calcular el rango intercuartiles usando los resultados de (1)

Escribamos una pequeña función que nos permita hacer este cálculo:

In [2]:
def calcular_iqr(data):
    # Calcular cuartiles 1 y 3:
    qs = data.quantile([0.25, 0.75])
    
    # Calcular IQR
    iqr = qs[0.75] - qs[0.25]
    
    return iqr

Y ahora sí usemos esta función para calcular el rango intercuartiles de la población y de la tasa de asesinatos. Calcularemos además la mediana en cada caso:

In [17]:
print('Población:')
print(f"\tMediana: {df['Population'].median()}")
print(f"\tIQR: {calcular_iqr(df['Population'])}")

print('='*30)
print('Tasa de asesinatos:')
print(f"\tMediana: {df['Murder.Rate'].median()}")
print(f"\tIQR: {calcular_iqr(df['Murder.Rate'])}")

Población:
	Mediana: 4436369.5
	IQR: 4847308.0
Tasa de asesinatos:
	Mediana: 4.0
	IQR: 3.125


Hemos caracterizado nuestras variables (población y tasa de asesinatos) con la media/desviación y con la mediana/IQR. Acá un resumen de los resultados obtenidos:

|  | Media | Desviación |  | Mediana | IQR |
|---|---|---|---|---|---|
| Población | 6'162.876 | 6'848.235 |  | 4'436.369 | 4'847.308 |
| Tasa de homicidios | 4.07 | 1.91 |  | 4.0 | 3.125 |


¿Cuál de las medidas es más confiable?

Vemos que en el caso de la población hay diferencias importantes entre media/desviación vs. mediana/IQR. Esto quiere decir que muy probablemente la variable población contiene valores extremos y por tanto la mediana y el IQR son más confiables.

En el caso de la tasa de homicidios vemos medidas comparables (con una desviación inferior que la IQR).

Así que si queremos generalizar la combinación mediana/IQR es más confiable que la media/desviación.

## 4. Resumen

|  | ¿Cuándo usar? | Limitaciones |
|---|---|---|
| Desviación estándar | Distribución normal y no hay presencia de *outliers* | Sensible a los *outliers* |
| Rango intercuartiles | Distribución no normal y cuando hay valores extremos |  |